text/2018-04-08-global-and-local-pinning.md

```diff
@@ -0,0 +1,247 @@
+
+The omission of `epoch::default_handle()` isn't a big problem since we already have `epoch::pin()` and `epoch::flush()`. Note that `epoch::pin()` is a much better option than `epoch::default_handle().pin()` anyway - it can automatically fall back to global pinning if the thread-local handle is destroyed:
+
+```rust
+fn pin() {
+    HANDLE.try_with(|h| h.pin()).unwrap_or_else(|| COLLECTOR.pin())
+}
+```
+
+Next, we implement `Send` for `Handle` and mark `Handle::pin()` as unsafe:
+
+```rust
+unsafe impl Send for Handle {}
+
+impl Handle {
+    unsafe fn pin(&self) -> Guard;
+}
+```
+
+The caller of `Handle::pin()` must pinky promise that the handle will not be sent to another thread while the `Guard` (or its clones) is alive. Unless, of course, the guard (and its clones) is simultaneously sent to the same thread, too.
+
+### Remove `is_pinned()`
+
+Another small tweak we'll do is remove `epoch::is_pinned()` and `Handle::is_pinned()`. These functions are only used by `crossbeam-deque` to make sure a fence is executed even if the thread is already pinned.
+
+Instead, we'll add functions `epoch::pin_fence()`, `Handle::pin_fence()`, and `Collector::pin_fence()`, which are equivalent to `pin()`, except they guarantee to execute a `SeqCst` fence.
+
+# Drawbacks
+
+* `epoch::default_handle()` is gone. But it doesn't seem to be useful and we have `epoch::pin()` and `epoch::flush()` anyway. We can reintroduce it in the future if a need actually comes up.
+
+* `Handle::clone()` is gone. However, nobody is using it and when it is used, it's used mistakenly. In short, it comes with practically zero benefits and a serious drawback.
+
+* `Handle::pin()` is now unsafe. See the *Alternatives* section for what we can do about it.
+
+# Alternatives
+
+### Add safe `Handle::pin_with()` and remove `Guard::clone()`
+
+We can add a safe alternative to `Guard::pin()`:
+
+```rust
+impl Handle {
+    fn pin_with<F, R>(&self, f: F) -> R
+    where
+        F: FnOnce(&mut Guard) -> R;
+}
+```
+
+This function is safe because it borrows the handle for the duration of pinning, which means it cannot be sent to another thread in the meantime. However, the user mustn't be able to clone the `Guard` and send the handle to another thread afterwards, as in this example:
+
+```rust
+let guard = handle.pin_with(|guard| guard.clone());
+thread::spawn(move || {
+    handle.pin_with(|_| ());
+});
+drop(guard); // Oops, the handle now belongs to another thread!
+```
+
+Therefore, we'd have to remove the `Clone` impl for `Guard`.
+
+### Make `Handle::pin()` safe, but add a lifetime to `Guard`
+
+Another alternative is to add a lifetime to `Guard` so that it becomes `Guard<'a>`. The lifetime `'a` borrows the `Handle` and therefore prevents it from being moved at all, which means it cannot be sent to another thread while such a guard exists.
+
+This way we don't have to remove `Handle::clone()`.
```

Comment:
I like this alternative, but if we do go with it, I suggest to remove `Handle::clone()` anyway, since as you mentioned it is not clear what the method should do.